File: server/policy.js
import { preferencesFooter } from './footer.js';
import { AI_BEHAVIORS, DEFAULT_POLICY, DEFAULT_PREFERENCES } from '../shared/features.js';

function invalid(message, status = 400) { throw Object.assign(new Error(message), { status }); }
function object(value) { return value && typeof value === 'object' && !Array.isArray(value); }

export function resolvePolicy(saved = {}) {
  return { ...DEFAULT_POLICY, ...saved, summarySchedule: { ...DEFAULT_POLICY.summarySchedule, ...saved.summarySchedule }, triggers: { ...DEFAULT_POLICY.triggers, ...saved.triggers }, behaviors: { ...DEFAULT_POLICY.behaviors, ...saved.behaviors }, folders: { ...DEFAULT_POLICY.folders, ...saved.folders }, content: { ...DEFAULT_POLICY.content, ...saved.content } };
}

export function updatePolicy(current, patch) {
  if (!object(patch)) invalid('AI permissions must be an object.');
  const next = resolvePolicy(current);
  for (const [key, value] of Object.entries(patch)) {
    if (!Object.hasOwn(DEFAULT_POLICY, key)) invalid(`Unknown permission: ${key}.`);
    if (key === 'enabled') {
      if (typeof value !== 'boolean') invalid('AI enabled must be true or false.');
      next.enabled = value;
    } else if (key === 'maxMessages') {
      if (!Number.isInteger(value) || value < 1 || value > 50) invalid('Context limit must be between 1 and 50 messages.');
      next.maxMessages = value;
    } else if (key === 'summarySchedule') {
      if (!object(value)) invalid('Summary schedule must be an object.');
      for (const [field, entry] of Object.entries(value)) {
        if (!Object.hasOwn(DEFAULT_POLICY.summarySchedule, field)) invalid('Unknown summary schedule setting.');
        if (field === 'cadence' && !['daily', 'interval'].includes(entry)) invalid('Choose daily or interval summaries.');
        if (field === 'time' && (typeof entry !== 'string' || !/^([01]\d|2[0-3]):[0-5]\d$/.test(entry))) invalid('Summary time must use HH:MM.');
        if (field === 'everyHours' && (!Number.isInteger(entry) || entry < 1 || entry > 168)) invalid('Summary interval must be 1–168 whole hours.');
        if (field === 'timeZone') {
          if (typeof entry !== 'string' || !entry.trim() || entry.length > 100) invalid('Choose a valid IANA time zone.');
          try { new Intl.DateTimeFormat('en', { timeZone: entry }).format(); } catch { invalid('Choose a valid IANA time zone, such as Asia/Hong_Kong.'); }
        }
        next.summarySchedule[field] = entry;
      }
    } else {
      if (!object(value)) invalid(`${key} must contain permission checkboxes.`);
      for (const [id, enabled] of Object.entries(value)) {
        if (!Object.hasOwn(DEFAULT_POLICY[key], id) || typeof enabled !== 'boolean') invalid(`Invalid ${key} permission: ${id}.`);
        next[key][id] = enabled;
      }
    }
  }
  return next;
}

export function updatePreferences(current, patch) {
  if (!object(patch)) invalid('Preferences must be an object.');
  const next = { ...DEFAULT_PREFERENCES, ...current };
  const choices = { signatureFormat: ['plain', 'html'], theme: ['system', 'light', 'dark'], density: ['comfortable', 'compact', 'spacious'], sort: ['newest', 'oldest', 'sender', 'subject', 'unread', 'starred'], replyTone: ['friendly', 'professional', 'concise', 'warm'], syncInterval: [0, 1, 5, 15, 30] };
  for (const [key, value] of Object.entries(patch)) {
    if (!Object.hasOwn(DEFAULT_PREFERENCES, key)) invalid(`Unknown preference: ${key}.`);
    if (choices[key]) {
      if (!choices[key].includes(value)) invalid(`Invalid ${key} preference.`);
    } else if (key === 'markReadOnOpen') {
      if (typeof value !== 'boolean') invalid('Mark read on open must be true or false.');
    } else {
      const max = key === 'signature' ? 12000 : ['language', 'translationLanguage'].includes(key) ? 60 : 100;
      if (typeof value !== 'string' || value.length > max || (key === 'language' && !value.trim())) invalid(`${key} must be text of at most ${max} characters.`);
      if (key === 'displayName' && /[\r\n]/.test(value)) invalid('Display name must be a single line.');
    }
    next[key] = value;
  }
  const footer = preferencesFooter(next);
  next.signature = next.signatureFormat === 'html' ? footer.html : footer.text;
  return next;
}

export function requireBehavior(policy, action) {
  const feature = AI_BEHAVIORS.find(item => item.id === action);
  if (!feature) invalid('Unknown AI behavior.');
  if (!policy.enabled || !policy.behaviors[action]) invalid(`${feature.label} is disabled in AI permissions.`, 403);
  const scope = { memory: 'contacts', research: 'contacts', meeting: 'calendar', schedule: 'calendar', attachments: 'attachments' }[action];
  if (scope && !policy.content[scope]) invalid(`Enable ${scope} access in AI permissions to use ${feature.label.toLowerCase()}.`, 403);
  if (action === 'batchReplies' && (!policy.content.sender || !policy.content.body)) invalid('Batch replies require sender and body access.', 403);
  return feature;
}

export function redactMessage(message, policy) {
  return {
    id: message.id, date: message.date, folder: message.folder, read: message.read, starred: message.starred,
    fromName: policy.content.sender ? message.fromName : '', fromEmail: policy.content.sender ? message.fromEmail : '',
    to: policy.content.sender ? message.to : '', subject: policy.content.subject ? message.subject : '',
    body: policy.content.body ? message.body : '', preview: policy.content.body ? message.preview : '',
    category: message.category, labels: policy.content.subject ? message.labels : [],
  };
}

export function permittedMessages(messages, policy) {
  return messages.filter(message => policy.folders[message.folder]).map(message => redactMessage(message, policy));
}

File: rust/src/policy.rs
use crate::{
    error::{Error, Result},
    store::{catalog, merge, string},
};
use serde_json::{Value, json};

pub fn resolve(saved: &Value) -> Value {
    let defaults = &catalog()["policy"];
    let mut value = merge(defaults.clone(), saved);
    for key in [
        "summarySchedule",
        "triggers",
        "behaviors",
        "folders",
        "content",
    ] {
        value[key] = merge(defaults[key].clone(), &saved[key]);
    }
    if saved["summarySchedule"].get("timeZone").is_none() {
        value["summarySchedule"]["timeZone"] = iana_time_zone::get_timezone()
            .unwrap_or_else(|_| "UTC".into())
            .into();
    }
    value
}
pub fn update(current: &Value, patch: &Value) -> Result<Value> {
    let patch = patch
        .as_object()
        .ok_or_else(|| Error::invalid("AI permissions must be an object."))?;
    let defaults = &catalog()["policy"];
    let mut next = resolve(current);
    for (key, value) in patch {
        if defaults.get(key).is_none() {
            return Err(Error::invalid("Unknown AI permission."));
        }
        match key.as_str() {
            "enabled" => {
                if !value.is_boolean() {
                    return Err(Error::invalid("AI enabled must be true or false."));
                }
                next[key] = value.clone();
            }
            "maxMessages" => {
                if !value.as_u64().is_some_and(|n| (1..=50).contains(&n)) {
                    return Err(Error::invalid(
                        "Context limit must be between 1 and 50 messages.",
                    ));
                }
                next[key] = value.clone();
            }
            "summarySchedule" => {
                let fields = value
                    .as_object()
                    .ok_or_else(|| Error::invalid("Summary schedule must be an object."))?;
                for (field, value) in fields {
                    let valid = match field.as_str() {
                        "cadence" => value
                            .as_str()
                            .is_some_and(|s| ["daily", "interval"].contains(&s)),
                        "time" => value.as_str().is_some_and(|s| {
                            s.len() == 5 && chrono::NaiveTime::parse_from_str(s, "%H:%M").is_ok()
                        }),
                        "everyHours" => value.as_u64().is_some_and(|n| (1..=168).contains(&n)),
                        "timeZone" => value
                            .as_str()
                            .is_some_and(|s| s.len() <= 100 && s.parse::<chrono_tz::Tz>().is_ok()),
                        _ => false,
                    };
                    if !valid {
                        return Err(Error::invalid(
                            "Choose a valid summary cadence, time, interval and IANA time zone.",
                        ));
                    }
                    next[key][field] = value.clone();
                }
            }
            _ => {
                let fields = value
                    .as_object()
                    .ok_or_else(|| Error::invalid("AI permissions must contain checkboxes."))?;
                for (field, value) in fields {
                    if defaults[key].get(field).is_none() || !value.is_boolean() {
                        return Err(Error::invalid("Invalid AI permission checkbox."));
                    }
                    next[key][field] = value.clone();
                }
            }
        }
    }
    Ok(next)
}
pub fn require(policy: &Value, action: &str) -> Result<Value> {
    let feature = catalog()["features"]
        .as_array()
        .unwrap()
        .iter()
        .find(|item| item["id"] == action)
        .ok_or_else(|| Error::invalid("Unknown AI behavior."))?;
    if policy["enabled"] != true || policy["behaviors"][action] != true {
        return Err(Error::new(
            403,
            "This behavior is disabled in AI permissions.",
        ));
    }
    let content = match action {
        "memory" | "research" => "contacts",
        "meeting" | "schedule" => "calendar",
        "attachments" => "attachments",
        _ => "",
    };
    if !content.is_empty() && policy["content"][content] != true {
        return Err(Error::new(
            403,
            "Enable the required content access in AI permissions.",
        ));
    }
    if action == "batchReplies"
        && (policy["content"]["sender"] != true || policy["content"]["body"] != true)
    {
        return Err(Error::new(
            403,
            "Batch replies require sender and body access.",
        ));
    }
    Ok(feature.clone())
}
pub fn redact(message: &Value, policy: &Value) -> Value {
    let mut result = json!({});
    for key in ["id", "date", "folder", "read", "starred", "category"] {
        if let Some(value) = message.get(key) {
            result[key] = value.clone();
        }
    }
    for (key, permission) in [
        ("fromName", "sender"),
        ("fromEmail", "sender"),
        ("to", "sender"),
        ("subject", "subject"),
        ("body", "body"),
        ("preview", "body"),
        ("labels", "subject"),
    ] {
        if policy["content"][permission] == true {
            if let Some(value) = message.get(key) {
                result[key] = value.clone();
            }
        } else {
            result[key] = if key == "labels" {
                json!([])
            } else {
                json!("")
            };
        }
    }
    result
}
pub fn matches_trigger(policy: &Value, trigger: &str, message: &Value) -> bool {
    let action = match trigger {
        "onOpen" | "onArrival" => "summary",
        "onReply" => "reply",
        _ => return false,
    };
    policy["enabled"] == true
        && policy["triggers"][trigger] == true
        && policy["behaviors"][action] == true
        && !message.is_null()
        && !["drafts", "trash"].contains(&string(message, "folder"))
        && policy["folders"][string(message, "folder")] == true
        && (policy["triggers"]["inboxOnly"] != true || message["folder"] == "inbox")
        && (policy["triggers"]["starredOnly"] != true || message["starred"] == true)
        && ["subject", "body", "sender"]
            .iter()
            .any(|field| policy["content"][field] == true)
}
